modin/pandas/test/internals/test_repartition.py

```diff
@@ -46,15 +47,22 @@ def test_repartition(axis, dtype):
 
     if dtype == "DataFrame":
         results = {
-            None: (1, 1),
-            0: (1, 2),
-            1: (2, 1),
+            None: ([4, 0, 0, 0], [3, 0, 0, 0]),
+            0: ([4, 0, 0, 0], [2, 1]),
+            1: ([2, 2], [3, 0, 0, 0]),
         }
     else:
         results = {
-            None: (1, 1),
-            0: (1, 1),
-            1: (2, 1),
+            None: ([4, 0, 0, 0], [1]),
+            0: ([4, 0, 0, 0], [1]),
+            1: ([2, 2], [1]),
         }
-
-    assert obj._query_compiler._modin_frame._partitions.shape == results[axis]
+    # in some cases empty partitions aren't filtered so it's better to use
+    # `row_lengths` and `column_widths` to check the result of repartitioning
+    # than `._partitions.shape`.
```

Comment:
IIUC empty partitions are not filtered out only in cases when there's not enough metadata to do so:
https://github.com/modin-project/modin/blob/f137564496ceff581a312de9af5ecbd143c7ee32/modin/core/dataframe/pandas/dataframe/dataframe.py#L471-L478
`._repartition` seems to have all the required metadata in the end, why does this change needed then?

Reply:
Previously `__make_init_labels_args` computed `_row_lengths_cache` and `_column_widths_cache`. Now this information is not available.

Reply:
Oh, I see.

But if we know the index we can precompute row/column lengths as well, right? We can reuse logic from `._reorder_labels.
https://github.com/modin-project/modin/blob/f137564496ceff581a312de9af5ecbd143c7ee32/modin/core/dataframe/pandas/dataframe/dataframe.py#L1114-L1119

Reply:
for example, in `broadcast_apply_full_axis` we can use the logic above if `keep_partitions=False` and `new_index/columns` is known.

Reply:
@dchigarev done, ready for review